src/executables/csql.c

```diff
@@ -1460,6 +1464,205 @@ csql_do_session_cmd (char *line_read, CSQL_ARGUMENT * csql_arg)
 	}
 
       break;
+
+    case S_CMD_CONNECT:
+      /*dbname_cpy can be stored DB name+ @ + remote hostname + \0 */
+      char dbname_cpy[MAXLOGNAME + CUB_MAXHOSTNAMELEN + 2];
+      char username_cpy[DB_MAX_USER_LENGTH + 1];
+      char host_cpy[DB_MAX_USER_LENGTH + 1];
+      const char *delim = " @\n";
+      char *save_ptr_strtok = NULL;
+      char *db_name = NULL;
+      char *user_name = NULL;
+      char *host_name = NULL;
+      char *p = NULL;
+      CSQL_ARGUMENT csql_new_arg;
+
+      /*find @hostname */
+      host_name = strchr (argument, '@');
+      if (host_name != NULL)
+	{
+	  if (*(host_name + 1) == '\0')
+	    {
+	      host_name = NULL;
+	      csql_Error_code = CSQL_ERR_CONNECT;
+	      return DO_CMD_FAILURE;
+	    }
+	  else
+	    {
+	      host_name += 1;
+	    }
+	}
+
+      if (host_name != NULL)
+	{
+	  if (strcasecmp ("127.0.0.1", host_name) == 0 || strcasecmp ("localhost", host_name) == 0
+	      || strcasecmp (boot_get_host_name (), host_name) == 0 || strcasecmp (boot_get_ip (), host_name) == 0)
+	    {
+	      *(host_name - 1) = '\0';
+	    }
+	}
+
+      strcpy (dbname_cpy, argument);
+      strcpy (username_cpy, argument);
+      /*find db name following the user name */
+      db_name = strrchr (dbname_cpy, ' ');
+      if (db_name != NULL)
+	{
+	  if (*(db_name + 1) == '\0')
+	    {
+	      db_name = NULL;
+	      csql_Error_code = CSQL_ERR_CONNECT;
+	      return DO_CMD_FAILURE;
+	    }
+	  else
+	    {
+	      db_name += 1;
+	    }
+	}
+      user_name = strtok_r (username_cpy, delim, &save_ptr_strtok);
+      if (user_name == NULL)
+	{
+	  csql_Error_code = CSQL_ERR_CONNECT;
+	  return DO_CMD_FAILURE;
+
+	}
+
+      memset (&csql_new_arg, 0, sizeof (CSQL_ARGUMENT));
+      csql_arg_copy (&csql_new_arg, csql_arg);
```

Comment:
`memcpy (&csql_new_arg, &csql_arg, sizeof (CSQL_ARGUMENT));`

And delete function **csql_arg_copy ()**

Reply:
It works, I will replace  csql_arg_copy() with memcpy(). Thank you.